ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -68,54 +74,61 @@
+        LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
+            stateStoreName, key);
+        throw new MaterializationException(String.format(
+            "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
+      }
 
-    LOG.debug("Handling pull query for key {} in partition {} of state store {}.",
-              key, metadata.partition(), stateStoreName);
-    
-    final HostInfo activeHost = metadata.activeHost();
-    final Set<HostInfo> standByHosts = metadata.standbyHosts();
-
-    // If the lookup is for a forwarded request, only filter localhost
-    List<KsqlHostInfo> allHosts = null;
-    if (routingOptions.skipForwardRequest()) {
-      LOG.debug("Before filtering: Local host {} ", localHost);
-      allHosts = ImmutableList.of(new KsqlHostInfo(localHost.getHost(), localHost.getPort()));
-    } else {
-      LOG.debug("Before filtering: Active host {} , standby hosts {}", activeHost, standByHosts);
-      allHosts = Stream.concat(Stream.of(activeHost), standByHosts.stream())
-          .map(this::asKsqlHost)
-          .collect(Collectors.toList());
+      LOG.debug("Handling pull query for key {} in partition {} of state store {}.",
+          key, metadata.partition(), stateStoreName);
+      final HostInfo activeHost = metadata.activeHost();
+      final Set<HostInfo> standByHosts = metadata.standbyHosts();
+
+      // If the lookup is for a forwarded request, only filter localhost
+      List<KsqlHostInfo> allHosts = null;
+      if (routingOptions.skipForwardRequest()) {
+        LOG.debug("Before filtering: Local host {} ", localHost);
+        allHosts = ImmutableList.of(new KsqlHostInfo(localHost.getHost(), localHost.getPort()));
+      } else {
+        LOG.debug("Before filtering: Active host {} , standby hosts {}", activeHost, standByHosts);
+        allHosts = Stream.concat(Stream.of(activeHost), standByHosts.stream())
+            .map(this::asKsqlHost)
+            .collect(Collectors.toList());
+      }
+      final RoutingFilter routingFilter = routingFilterFactory.createRoutingFilter(routingOptions,
+          allHosts, activeHost, applicationId, stateStoreName, metadata.partition());
+
+      // Filter out hosts based on active, liveness and max lag filters.
+      // The list is ordered by routing preference: active node is first, then standby nodes.
+      // If heartbeat is not enabled, all hosts are considered alive.
+      // If the request is forwarded internally from another ksql server, only the max lag filter
+      // is applied.
+      final ImmutableList<KsqlNode> filteredHosts = allHosts.stream()
+          .filter(routingFilter::filter)
+          .map(this::asNode)
+          .collect(ImmutableList.toImmutableList());
+
+      LOG.debug("Filtered and ordered hosts: {}", filteredHosts);
+
+      groups.computeIfAbsent(filteredHosts, nl -> new ArrayList<>());
+      groups.get(filteredHosts).add(key);
```

Comment:
nit: `computeIfAbsent` returns the result of the compute, so you could just 
```suggestion
      groups.computeIfAbsent(filteredHosts, nl -> new ArrayList<>()).add(key);
```